NeoML/src/Dnn/Layers/LoraFullyConnectedLayer.cpp

```diff
@@ -0,0 +1,316 @@
+#include <NeoML/Dnn/Layers/DropoutLayer.h>
+#include <NeoML/Dnn/Layers/EltwiseLayer.h>
+#include <NeoMathEngine/NeoMathEngine.h>
+
+namespace NeoML {
+
+static const char* const baseFcName = "FullyConnectedBase";
+static const char* const loraFcAName = "FullyConnectedALoRA";
+static const char* const loraFcBName = "FullyConnectedBLoRA";
+static const char* const loraDropoutName = "DropoutLoRA";
+static const char* const loraScalingName = "ScalingLoRA";
+static const char* const loraSumName = "SumLoRA";
+
+//----------------------------------------------------------------------------------------------
+
+CLoraFullyConnectedLayer::CLoraFullyConnectedLayer( IMathEngine& mathEngine, const char* name ) :
+	CCompositeLayer( mathEngine, ( name == nullptr ) ? "CLoraFullyConnectedLayer" : name )
+{
+	buildLayer();
+}
+
+CLoraFullyConnectedLayer::CLoraFullyConnectedLayer( CFullyConnectedLayer& fc ) :
+	CCompositeLayer( fc.MathEngine(), fc.GetName() ),
+	baseFc( &fc )
+{
+	baseFc->SetName( baseFcName );
+	SetInputMapping( *baseFc );
+	SetOutputMapping( *baseFc );
+	AddLayer( *baseFc );
+
+	NeoPresume( baseFc != nullptr && baseFc->GetName() == CString( baseFcName ) );
+}
+
+static const int LoraFullyConnectedLayerVersion = 0;
+
+void CLoraFullyConnectedLayer::Serialize( CArchive& archive )
+{
+	archive.SerializeVersion( LoraFullyConnectedLayerVersion );
+	CCompositeLayer::Serialize( archive );
+
+	archive.Serialize( loraRank );
+	archive.Serialize( loraAlpha );
+	archive.Serialize( loraDropoutRate );
+
+	if( archive.IsLoading() ) {
+		baseFc = CheckCast<CFullyConnectedLayer>( GetLayer( baseFcName ) );
+		// optional LoRA layers
+		loraFcA = HasLayer( loraFcAName )
+			? CheckCast<CFullyConnectedLayer>( GetLayer( loraFcAName ) )
+			: nullptr;
+
+		if( loraFcA != nullptr ) {
+			NeoAssert( loraRank != 0 && loraAlpha != 0 );
+
+			loraFcB = CheckCast<CFullyConnectedLayer>( GetLayer( loraFcBName ) );
+			loraDropout = HasLayer( loraDropoutName )
+				? CheckCast<CDropoutLayer>( GetLayer( loraDropoutName ) )
+				: nullptr;
+			NeoAssert( loraDropout == nullptr || loraDropoutRate > 0 );
+			loraScaling = CheckCast<CLinearLayer>( GetLayer( loraScalingName ) );
+			loraSum = CheckCast<CEltwiseSumLayer>( GetLayer( loraSumName ) );
+		} else {
+			NeoAssert( loraRank == 0 && loraAlpha == 0 && loraDropoutRate == 0 );
+			// Otherwise, it isn't built
+			loraFcB = nullptr;
+			loraDropout = nullptr;
+			loraScaling = nullptr;
+			loraSum = nullptr;
+		}
+	}
+}
+
+void CLoraFullyConnectedLayer::buildLayer()
+{
+	baseFc = FINE_DEBUG_NEW CFullyConnectedLayer( MathEngine() );
+	baseFc->SetName( baseFcName );
+	baseFc->SetNumberOfElements( 1 );
+	SetInputMapping( *baseFc );
+	SetOutputMapping( *baseFc );
+	AddLayer( *baseFc );
+
+	NeoPresume( baseFc != nullptr && baseFc->GetName() == CString( baseFcName ) );
+}
+
+void CLoraFullyConnectedLayer::Reshape()
+{
+	CheckLayerArchitecture( GetInputCount() == 1, "Layer must have only 1 input" );
+	CCompositeLayer::Reshape();
```

Comment:
You can get rid of `initedFirst` by calculating
`const bool hasUinitializedLora = ( loraFcA != nullptr && loraFcA->Weights() == nullptr );` on the fly.

The key point is to store this flag before `CCompositeLayer::Reshape()` and initialize the weights after base class `Reshape()`

Reply:
Thank you, it is important fix. I did not serialize `initedFirst`, so it may cause problems in some cases.
Fixed!